include/ensmallen_bits/demon_adam/demon_adam_update.hpp

```diff
@@ -0,0 +1,169 @@
+/**
+ * @file demon_sgd_update.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of DemonAdam.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_DEMON_ADAM_DEMON_ADAM_UPDATE_HPP
+#define ENSMALLEN_DEMON_ADAM_DEMON_ADAM_UPDATE_HPP
+
+#include <assert.h>
+
+namespace ens {
+
+/**
+ * DemonAdam automatically decays momentum, motivated by decaying the total
+ * contribution of a gradient to all future updates. 
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @misc{
+ *   title   = {Decaying momentum helps neural network training},
+ *   author  = {John Chen and Cameron Wolfe and Zhao Li
+ *              and Anastasios Kyrillidis},
+ *   url     = {https://arxiv.org/abs/1910.04952}
+ *   year    = {2019}
+ * }
+ * @endcode
+ *
+ * @tparam UpdateRule DemonAdam optimizer update rule to be used.
+ */
+template<typename UpdateRule>
+class DemonAdamUpdate
+{
+ public:
+  /**
+   * Construct the DemonAdam update policy with the given parameters.
+   *
+   * @param momentumIterations The number of iterations before the momentum
+   *     will decay to zero.
+   * @param momentum The initial momentum coefficient.
+   * @param adamUpdate Instantiated Adam update policy used to adjust the given
+   *     parameters.
+   */
+  DemonAdamUpdate(const size_t momentumIterations = 100,
+                  const double momentum = 0.9,
+                  const UpdateRule& adamUpdate = UpdateRule()) :
+      T(momentumIterations),
+      betaInit(momentum),
+      t(0),
+      adamUpdateInst(adamUpdate)
+  {
+    // Make sure the momentum iterations parameter is non-zero.
+    assert(momentumIterations != 0 && "The number of iterations before the "
+        "momentum will decay is zero, make sure the max iterations and "
+        "batch size parameter is set correctly. "
+        "Default: momentumIterations = maxIterations * batchSize.");
```

Comment:
Shouldn't this be `maxIterations / batchSize` as the default?  Otherwise the number of decay steps for momentum won't match the number of steps taken during optimization.

Reply:
Nice catch!